Document/0x05e-Testing-Cryptography.md

```diff
@@ -232,3 +232,86 @@ public static void main (String args[]) {
 
 ##### CWE
 - CWE-330: Use of Insufficiently Random Values
+
+
+### Testing Key Management
+
+#### Overview
+
+Symmetric cryptography provides confidentiality and integrity of data because it ensures one basic cryptographic principle. It is based on the fact that a given ciphertext can only, in any circumstance, be decrypted when providing the original encryption key. The security problem is thereby shifted to securing the key instead of the content that is now securely encrypted. Asymmetric cryptography solves this problem by introducing the concept of a private and public key pair. The public key can be distributed freely, the private key is kept secret.
+
+When testing an Android application on correct usage of cryptography it is thereby also important to make sure that key material is securely generated and stored. In this section we will discuss different ways to manage cryptographic keys and how to test for them. We discuss the most secure way, down to the less secure way of generating and storing key material.
+
+The most secure way of handling key material, is simply never storing it on the filesystem. This means that the user should be prompted to input a passphrase every time the application needs to perform a cryptographic operation. Although this is not the ideal implementation from a user experience point of view, it is however the most secure way of handling key material. The reason is because key material will only be available in an array in memory while it is being used. Once the key is not needed anymore, the array can be zeroed out. This minimizes the attack window as good as possible. No key material touches the filesystem and no passphrase is stored.  However, note that some ciphers do not properly clean up their byte-arrays. For instance, the AES Cipher in BouncyCastle does not always clean up its latest working key.
+
+The encryption key can be generated from the passphrase by using the Password Based Key Derivation Function version 2 (PBKDFv2). This cryptographic protocol is designed to generate secure and non brute-forceable keys. The code listing below illustrates how to generate a strong encryption key based on a password.
+
+```Java
+public static SecretKey generateStrongAESKey(char[] password, int keyLength)
+{
+    //Initiliaze objects and variables for later use
+    int iterationCount = 10000;
+    int saltLength     = keyLength / 8;
+    SecureRandom random = new SecureRandom();
+    
+    //Generate the salt
+    byte[] salt = new byte[saltLength];
+    randomb.nextBytes(salt);
```

Comment:
A salt is usually used to counter a pre-computated attack of encyption keys or hashes also known as a rainbow table attack. An attacker pre-computes all the different possibilities and stores them in a big table. These tables can easily reach several gigabytes in size. When an attacker then wants to bruteforce  a key or hash he simply iterates over the table of pre-computated hashses or  keys and tries them one after one. He does not need to do the work of generating or computing the keys/hashes anymore. By adding a salt to the key or hash generation phase you basically render these rainbow tables useless. An attacker will need to use the hash and re-compute all the different hashes/keys in a brute-force attack. Because the PBKDFv2 is specifically designed to slow down the generation process (by increasing the iteration count) this will significantly increase the effort for a brute force attack.